### ISSUE TITLE
Shared components need to be able to import client components

### ISSUE THREAD

**Issue by gaearon** [vercel/next.js#34594]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000
    Binaries:
      Node: 17.5.0
      npm: 8.4.1
      Yarn: 1.22.15
      pnpm: 6.11.0
    Relevant packages:
      next: 12.1.1-canary.1
      react: 0.0.0-experimental-13036bfbc-20220121
      react-dom: 0.0.0-experimental-13036bfbc-20220121

### Describe the Bug

Importing a Client Component from a Shared (via Server) component runs it in the Server environment. But it should turn it into a Module Reference instead.

So `index.server.js` -> `components.js` -> `thing.client.js` tries to run `<Thing />` *in the SC environment*. But it should treat it as a client import, and "skip" over it.

https://github.com/gaearon/sc-bug-repro/blob/df693017248e13c2c59829a27aa7ff6b63aad341/pages/index.server.js#L3-L8

In other words, imports to Client files only seem to be replaced with Module References inside `.server.js` files. But they should be replaced with Module References in all files.

### Expected Behavior

It should work

### To Reproduce

```
git clone https://github.com/gaearon/sc-bug-repro.git
cd sc-bug-repro

git checkout shared-client

npm install
npm run dev
```

See `pages/index.server.js` for two snippets that should be equivalent but aren't.
